File: pod_api/bin/app/models/chart_item.dart
import 'package:xml/xml.dart' as xml;
import 'package:meta/meta.dart';

import 'lookup_item.dart';

class ChartItem {
  final String id;
  final String title;
  final String summary;
  final String artist;
  final String thumbnail;
  final String releaseDate;
  final String category;
  LookupResult lookup;

  ChartItem({
    @required this.id,
    @required this.title,
    @required this.summary,
    @required this.artist,
    @required this.thumbnail,
    @required this.releaseDate,
    @required this.category,
    this.lookup,
  });

  static ChartItem fromXml(xml.XmlElement xmlNode) {
    var id = xmlNode.findAllElements('id').single.getAttribute('im:id');
    var title = xmlNode.findAllElements('title').single.text;
    var summary = xmlNode.findAllElements('summary').single.text;
    var artist = xmlNode.findAllElements('im:artist').single.text;
    var thumbnail = xmlNode.findAllElements('im:image').last.text;
    var releaseDate = xmlNode.findAllElements('im:releaseDate').single.text;
    var category = xmlNode.findAllElements('category').single.getAttribute('label');

    return ChartItem(
      id: id,
      title: title,
      summary: summary,
      artist: artist,
      thumbnail: thumbnail,
      releaseDate: releaseDate,
      category: category,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'title': title,
      'summary': summary,
      'artist': artist,
      'thumbnail': thumbnail,
      'releaseDate': releaseDate,
      'category': category,
      'lookup': this.lookup?.toMap() ?? {},
    };
  }
}

File: podapi_ts/src/Models/ChartItem.ts
import { JsonIgnore, JsonProperty } from "json-object-mapper";
import StringMap from "../Interfaces/StringMapInterface";
import EntryDTO from "./DTOs/EntryDTO";
import { ILookupItem } from "./LookupItem";
import Model from "./Model";

// tslint:disable-next-line:interface-name
export default interface IChartItem {
  id: string;
  title: string;
  summary: string;
  artist: string;
  thumbnails: string[];
  releaseDate: string;
  category: string;
  lookup: ILookupItem;
}

export class ChartItem extends Model<EntryDTO> implements IChartItem {
  @JsonProperty()
  public id: string;

  @JsonProperty()
  public title: string;

  @JsonProperty()
  public summary: string;

  @JsonProperty()
  public artist: string;

  @JsonProperty()
  public thumbnails: string[];

  @JsonProperty()
  public releaseDate: string;

  @JsonProperty()
  public category: string;

  @JsonProperty()
  public lookup: ILookupItem;

  constructor(dto: EntryDTO) {
    super(dto);
  }

  public mapFromDTO(dto: EntryDTO): void {
    this.id = dto.id.id;
    this.title = dto.title;
    this.summary = dto.summary;
    this.thumbnails = dto.image.map(imageDTO => imageDTO.text);
    this.releaseDate = dto.releaseDate.label;
    this.category = dto.category.label;

    if (typeof dto.artist === "object") {
      this.artist = ((dto.artist as unknown) as StringMap)["#text"];
    } else if (typeof dto.artist === "string") {
      this.artist = dto.artist as string;
    }
  }
}
